capsules/src/seven_segment.rs

```diff
@@ -0,0 +1,371 @@
+//!         segment_array,
+//!         digit_array,
+//!         buffer,
+//!         virtual_alarm_digit,
+//!         kernel::hil::gpio::ActivationMode::ActiveLow,
+//!         kernel::hil::gpio::ActivationMode::ActiveHigh,
+//!         60
+//!     ),
+//! );
+//!
+//! virtual_alarm_digit.set_alarm_client(digit_display);
+//!
+//! digit_display.init();
+//!
+//! Syscall Interface
+//! -----------------
+//!
+//! ### Command
+//!
+//! All operations are synchronous, so this capsule only uses the `command`
+//! syscall.
+//!
+//! #### `command_num`
+//!
+//! - `0`: Return the number of digits on the display being used.
+//!   - `data1`: Unused.
+//!   - `data2`: Unused.
+//!   - Return: Number of digits.
+//! - `1`: Prints one digit at the requested position.
+//!   - `data1`: The position of the digit. Starts at 1.
+//!   - `data2`: The digit to be represented, from 0 to 9.
+//!   - Return: `Ok(())` if the digit index was valid, `INVAL` otherwise.
+//! - `2`: Clears all digits currently being displayed.
+//!   - `data1`: Unused.
+//!   - `data2`: Unused.
+//! - `3`: Print a dot at the requested digit position.
+//!   - `data1`: The position of the dot. Starts at 1.
+//!   - Return: `Ok(())` if the index was valid, `INVAL` otherwise.
+//! - `4`: Print a custom pattern for a digit on a certain position.
+//!   - `data1`: The position of the digit. Starts at 1.
+//!   - `data2`: The custom pattern to be represented.
+//!   - Return: `Ok(())` if the index was valid, `INVAL` otherwise.
+
+use core::cell::Cell;
+
+use kernel::hil::gpio::{ActivationMode, Pin};
+use kernel::hil::time::{Alarm, AlarmClient, ConvertTicks};
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::utilities::cells::TakeCell;
+use kernel::ErrorCode;
+use kernel::ProcessId;
+
+/// Syscall driver number.
+use crate::driver;
+pub const DRIVER_NUM: usize = driver::NUM::SevenSegment as usize;
+
+/// Digit patterns
+//
+//      A
+//      _
+//   F |_| B       center = G
+//   E |_| C . Dp
+//      D
+//
+const DIGITS: [u8; 10] = [
+    // pattern: 0bDpGFEDCBA
+    0b00111111, // 0
+    0b00000110, // 1
+    0b01011011, // 2
+    0b01001111, // 3
+    0b01100110, // 4
+    0b01101101, // 5
+    0b01111101, // 6
+    0b00100111, // 7
+    0b01111111, // 8
+    0b01101111, // 9
+];
+
+/// Holds an array of digits and an array of segments for each digit.
+
+pub struct SevenSegmentDriver<'a, P: Pin, A: Alarm<'a>, const NUM_DIGITS: usize> {
+    /// An array of 8 segments (7 for digit segments and one dot segment)
+    segments: &'a [&'a P; 8],
+    /// An array of 4 digits
+    digits: &'a [&'a P; NUM_DIGITS],
+    /// A buffer which contains the patterns displayed for each digit
+    /// Each element of the buffer array represents the pattern for one digit, and
+    /// is a sequence of bits that have the value 1 for a lit segment and the value 0 for
+    /// an unlit segment.
+    buffer: TakeCell<'a, [u8; NUM_DIGITS]>,
+    alarm: &'a A,
```

Comment:
What is the alarm for? I don't think we can have a purely synchronous interface and use an alarm.

Reply:
The alarm is for the refresh rate of the display. Since the digits and the segments are multiplexed, not all digits can be represented simultaneously, so I used the alarm to switch between digits after a short amount of time, to appear as if all digits are being shown at the same instance.